### ISSUE TITLE
Placeholder for old PR #529

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#529]
Originally raised by @verdverm in https://github.com/cuelang/cue/pull/529

**Comment by cueckoo** [comment]
_Original reply by @googlebot in https://github.com/cuelang/cue/pull/529#issuecomment-703380178_

All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.

We need to confirm that all authors are ok with their commits being contributed to this project.  Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.

*Note to project maintainer:* There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent.  In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).

ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%!A(MISSING)%!F(MISSING)%!F(MISSING)github.com%!F(MISSING)cuelang%!F(MISSING)cue%!F(MISSING)pull%!F(MISSING)529) for more info**.

**Comment by cueckoo** [comment]
_Original reply by @googlebot in https://github.com/cuelang/cue/pull/529#issuecomment-703380181_

All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.

We need to confirm that all authors are ok with their commits being contributed to this project.  Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.

*Note to project maintainer:* There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent.  In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).

ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%!A(MISSING)%!F(MISSING)%!F(MISSING)github.com%!F(MISSING)cuelang%!F(MISSING)cue%!F(MISSING)pull%!F(MISSING)529) for more info**.

**Comment by cueckoo** [comment]
_Original reply by @verdverm in https://github.com/cuelang/cue/pull/529#issuecomment-703380442_

@mpvl did I mess this PR up with a rebase?

I addressed the comments you made but now I see a ton of other commits... 

---

(edit) think I fixed the rebase issue 🤦

**Comment by cueckoo** [comment]
_Original reply by @googlebot in https://github.com/cuelang/cue/pull/529#issuecomment-703380902_

CLAs look good, thanks!

ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%!A(MISSING)%!F(MISSING)%!F(MISSING)github.com%!F(MISSING)cuelang%!F(MISSING)cue%!F(MISSING)pull%!F(MISSING)529) for more info**.

**Comment by cueckoo** [comment]
_Original reply by @googlebot in https://github.com/cuelang/cue/pull/529#issuecomment-703380903_

CLAs look good, thanks!

ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%!A(MISSING)%!F(MISSING)%!F(MISSING)github.com%!F(MISSING)cuelang%!F(MISSING)cue%!F(MISSING)pull%!F(MISSING)529) for more info**.

**Comment by cueckoo** [comment]
_Original reply by @verdverm in https://github.com/cuelang/cue/pull/529#issuecomment-712358959_

I've found the map to be less ideal, preferring to have the written order maintained, like Cue does elsewhere.

Now thinking a `Attrs.Values() cue.Value` with an underlying struct might make more sense.

Thoughts?

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/pull/529#issuecomment-768049051_

Note that there are now also attribute at the declaration level. So the API needs to take these into account as well.

**Comment by cueckoo** [comment]
_Original reply by @verdverm in https://github.com/cuelang/cue/pull/529#issuecomment-768057530_

tbh, the Attribute API is the more restrictive one for me than the `Value.Attribute()` vs getting all of them.

Are they both using the same underlying Attribute type?

I'd like to be able to extract all of the content from within the Attribute `()` without having to parse it myself

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/pull/529#issuecomment-769306834_

I've added a function to get declaration attributes. This may help in coming up with an API that supports both. 

https://cue-review.googlesource.com/c/cue/+/8342

**Comment by cueckoo** [comment]
_Original reply by @verdverm in https://github.com/cuelang/cue/pull/529#issuecomment-780800093_

@mpvl this seems to start supporting the suggestion you made here: https://github.com/cuelang/cue/issues/774#issuecomment-780370962

I have a number of open issues / requests / etc around Attributes, maybe a proposal of sorts is in order for them? I can start a new issue outlining all of these, something like "Attribute(s) API"

thoughts?

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/pull/529#issuecomment-781432543_

@verdverm : I've started converging on what would be a proper API, so maybe we don't need to go through these lengths.

In a nutshell, we ideally should have only _one_ call for fetching attributes (`Value.Attributes` or `Value.Attrs`) that returns _all_ attributes associated with a value, so field and declaration attributes, but then annotate the kind in the `Attribute` struct. This would allow users to easily interpret declaration attrs of a field as if it were field attrs, etc., without having to jump through hoops and making multiple calls. It also keeps the API tight.

In order to keep the API returning stable when we later introduce more attribute types, this method could take a mask specifying the kind of desired attrs:
```
type AttrKind struct
const (
    FieldAttr AttrKind = 1 << iota
    DeclAttr
    // Possible future attr kinds
    // ElemAttr
    // FileAttr
    // ValueAttr = FieldAttr|DeclAttr|ElemAttr
) 
```

The `Attribute` type could then have a `Kind` method returning this type and `Value.Attributes(mask AttrKind)` would take a mask to only return the requested attributes.

This is not too far off from the implementation in this CL. The heavy lifting of extracting the attrs has been implemented in `internal/core/export`, so this should be fairly straightforward.

The `Attribute` type should have some lower-level non-opinionated functionality, like `Body()` or `Contents()` or something, in addition to the `Name()` method you added (should it be `Name` or `Key`? Either way I guess).

Does this sound reasonable?
